### ISSUE TITLE
calico-node on Kubernetes 'forgets' about IPinIP encapsulation with Kubernetes API unavailable

### ISSUE THREAD

**Issue by embik** [projectcalico/calico#2522]
When the Kubernetes API is unavailable, even for a short period of time, the Calico DaemonSet 'forgets' about IPinIP encapsulation for pod networking and tries to send packages directly via eth0 without encapsulation. This is likely filtered by anti-spoofing measurements on our infrastructure. This causes Kubernetes pods on other nodes to be unavailable during Kubernetes API downtime.

My understanding is that IPinIP is configured in the IPPool resource which is stored in the Kubernetes API. If the datastore and therefore the resource is unavailable, Calico seems to drop the present IPinIP configuration.

## Expected Behavior
Calico remembers that IPinIP is configured for the Kubernetes pod network and uses IPinIP encapsulation to reach pods on other nodes known before the Kubernetes API downtime (caches known endpoints in some way).

## Current Behavior
When the Kubernetes API goes down, Calico reconfigures and disables IPinIP via the `tunl0` interface. Looking at `/etc/calico` within the calico container this changes:
```diff
diff -r calico-api-up/confd/config/bird_ipam.cfg calico-api-down/confd/config/bird_ipam.cfg
13c13,22
< filter calico_kernel_programming {
< 
<   if ( net ~ 10.244.0.0/16 ) then {
< 
<     krt_tunnel = "tunl0";         
<     accept;
<   }  
< 
<   accept;                                  
< }
---
> filter calico_kernel_programming { accept; }
```

Changes to `ip route` output (excerpt):
```diff
< 10.244.1.2 via 10.0.5.19 dev tunl0 proto bird onlink 
< 10.244.1.4 via 10.0.5.19 dev tunl0 proto bird onlink 
< 10.244.1.8 via 10.0.5.19 dev tunl0 proto bird onlink 
> 10.244.1.2 via 10.0.5.19 dev eth0 proto bird 
> 10.244.1.4 via 10.0.5.19 dev eth0 proto bird 
> 10.244.1.8 via 10.0.5.19 dev eth0 proto bird
```

`tcpdump` on `eth0` when the API is available (source pod has ip `10.244.2.56`, target pod has ip `10.244.3.227`): IPinIP is used.
```
07:02:09.014240 IP (tos 0x0, ttl 63, id 30437, offset 0, flags [DF], proto IPIP (4), length 80)
    10.0.5.15 > 10.0.5.4: IP (tos 0x0, ttl 63, id 10438, offset 0, flags [DF], proto TCP (6), length 60)
    10.244.2.56.52176 > 10.244.3.227.8080: Flags [S], cksum 0x0acc (correct), seq 2851428316, win 27200, options [mss 1360,sackOK,TS val 1272151631 ecr 0,nop,wscale 7], length 0
07:02:09.015229 IP (tos 0x0, ttl 63, id 30438, offset 0, flags [DF], proto IPIP (4), length 72)
    10.0.5.15 > 10.0.5.4: IP (tos 0x0, ttl 63, id 10439, offset 0, flags [DF], proto TCP (6), length 52)
    10.244.2.56.52176 > 10.244.3.227.8080: Flags [.], cksum 0x6f91 (correct), seq 2851428317, ack 1434953984, win 213, options [nop,nop,TS val 1272151632 ecr 3282727114], length 0
07:02:09.015351 IP (tos 0x0, ttl 63, id 30439, offset 0, flags [DF], proto IPIP (4), length 153)
    10.0.5.15 > 10.0.5.4: IP (tos 0x0, ttl 63, id 10440, offset 0, flags [DF], proto TCP (6), length 133)
    10.244.2.56.52176 > 10.244.3.227.8080: Flags [P.], cksum 0x7daf (correct), seq 0:81, ack 1, win 213, options [nop,nop,TS val 1272151632 ecr 3282727114], length 81: HTTP, length: 81
	GET / HTTP/1.1
	Host: 10.244.3.227:8080
	User-Agent: curl/7.61.0
	Accept: */*
	
07:02:09.016502 IP (tos 0x0, ttl 63, id 30440, offset 0, flags [DF], proto IPIP (4), length 72)
    10.0.5.15 > 10.0.5.4: IP (tos 0x0, ttl 63, id 10441, offset 0, flags [DF], proto TCP (6), length 52)
    10.244.2.56.52176 > 10.244.3.227.8080: Flags [.], cksum 0x6ea5 (correct), seq 81, ack 146, win 221, options [nop,nop,TS val 1272151633 ecr 3282727115], length 0
07:02:09.017187 IP (tos 0x0, ttl 63, id 30441, offset 0, flags [DF], proto IPIP (4), length 72)
    10.0.5.15 > 10.0.5.4: IP (tos 0x0, ttl 63, id 10442, offset 0, flags [DF], proto TCP (6), length 52)
    10.244.2.56.52176 > 10.244.3.227.8080: Flags [F.], cksum 0x6ea3 (correct), seq 81, ack 146, win 221, options [nop,nop,TS val 1272151634 ecr 3282727115], length 0
07:02:09.017790 IP (tos 0x0, ttl 63, id 30442, offset 0, flags [DF], proto IPIP (4), length 72)
    10.0.5.15 > 10.0.5.4: IP (tos 0x0, ttl 63, id 10443, offset 0, flags [DF], proto TCP (6), length 52)
    10.244.2.56.52176 > 10.244.3.227.8080: Flags [.], cksum 0x6ea0 (correct), seq 82, ack 147, win 221, options [nop,nop,TS val 1272151635 ecr 3282727116], length 0
```

`tcpdump` on `eth0` when the API is unavailable: IPinIP is not used.
```
07:06:30.904165 IP (tos 0x0, ttl 63, id 52994, offset 0, flags [DF], proto TCP (6), length 60)
    10.244.2.56.56116 > 10.244.3.227.8080: Flags [S], cksum 0x1c31 (incorrect -> 0x32ca), seq 441495838, win 27200, options [mss 1360,sackOK,TS val 1272413516 ecr 0,nop,wscale 7], length 0
07:06:31.973562 IP (tos 0x0, ttl 63, id 52995, offset 0, flags [DF], proto TCP (6), length 60)
    10.244.2.56.56116 > 10.244.3.227.8080: Flags [S], cksum 0x1c31 (incorrect -> 0x2ea6), seq 441495838, win 27200, options [mss 1360,sackOK,TS val 1272414576 ecr 0,nop,wscale 7], length 0
07:06:34.012066 IP (tos 0x0, ttl 63, id 52996, offset 0, flags [DF], proto TCP (6), length 60)
    10.244.2.56.56116 > 10.244.3.227.8080: Flags [S], cksum 0x1c31 (incorrect -> 0x26a7), seq 441495838, win 27200, options [mss 1360,sackOK,TS val 1272416623 ecr 0,nop,wscale 7], length 0
07:06:38.043774 IP (tos 0x0, ttl 63, id 52997, offset 0, flags [DF], proto TCP (6), length 60)
    10.244.2.56.56116 > 10.244.3.227.8080: Flags [S], cksum 0x1c31 (incorrect -> 0x16e7), seq 441495838, win 27200, options [mss 1360,sackOK,TS val 1272420655 ecr 0,nop,wscale 7], length 0
07:06:46.235747 IP (tos 0x0, ttl 63, id 52998, offset 0, flags [DF], proto TCP (6), length 60)
    10.244.2.56.56116 > 10.244.3.227.8080: Flags [S], cksum 0x1c31 (incorrect -> 0xf6e6), seq 441495838, win 27200, options [mss 1360,sackOK,TS val 1272428847 ecr 0,nop,wscale 7], length 0
```

## Possible Solution

Not sure here, but maybe prevent calico from updating `confd/config/bird_ipam.cfg` in case of Kubernetes API unavailability.

## Steps to Reproduce (for bugs)
1. Install Calico as Kubernetes Network and Policy Addon
2. Check IPinIP encapsulation and config file (should be encapsulated)
3.  Shut down Kubernetes API
4. Check IPinIP encapsulation and config file (should not be encapsulated)

## Context
We are running the Kubernetes API without a HA setup and sometimes need to take down the API for maintenance tasks (e.g. updating to a new Kubernetes release). We'd like the running workloads on the cluster (e.g. the pod connectivity) to be not affected by this.

## Your Environment

* Calico version: 3.6.1
* Orchestrator version (e.g. kubernetes, mesos, rkt): Kubernetes v1.12.7
* Operating System and version: CoreOS 2023.5.0

**Comment by fasaxc** [comment]
@embik Is there anything in the calico/node logs at the time when the config gets regenerated?

**Comment by embik** [comment]
Here's a log file up to the message of `bird_ipam.cfg` being updated (it's the last occurence of updates to this file). A few duplicate lines (`failed to create watcher` and GOAWAY messages) have been removed for readability. I don't see any log entries that indicate why it drops the `tunl0` interface as target (I suspect that's what `bird_ipam.cfg` configures, basically) for the pod network.

[calico.log](https://github.com/projectcalico/calico/files/3037657/calico.log)

We added this Felix environment variable in hope it would force IPinIP use, but there's no change here. Just in case some logs seem strange here.
```yaml
- name: FELIX_IPINIPENABLED
   value: "true"
```

**Comment by embik** [comment]
Hey @fasaxc @caseydavenport, just to ensure I'm a good citizen here, do you need any more information regarding this? Have you been able to reproduce this behaviour? If not, I could invest some time to create more detailed steps to reproduce.

**Comment by caseydavenport** [comment]
@embik I think we have enough for now, but will report back. I haven't had a chance to reproduce yet.

Just to confirm, this is using the "Kubernetes API datastore", not etcd?

**Comment by embik** [comment]
@caseydavenport Yes, it's the Kubernetes API datastore (source of my installed manifest](https://docs.projectcalico.org/v3.6/getting-started/kubernetes/installation/calico#installing-with-the-kubernetes-api-datastore50-nodes-or-less)).

**Comment by caseydavenport** [comment]
Quick update here - @jpalardy has done some investigation into this and we think this was already fixed in `master` by this PR: https://github.com/projectcalico/libcalico-go/pull/1057

@jpalardy is working on backporting this fix so we can include in the next bugfix release.

**Comment by embik** [comment]
Thanks for the update, in that case I'm looking forward to the next bugfix release :-)

**Comment by caseydavenport** [comment]
Going to close this now since the fix has been merged to release-v3.5 and release-v3.6, and is included in the newly minted v3.7.0.

Will be available in v3.6.2 and v3.5.5 when they are cut.
